File: 2.两数相加/TowNumberAdd.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 * C语言解答版本
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    int value1[100] = {0};
    int value2[100] = {0};
    int value3[100] = {0};
    int length1 = -1, length2 = -1, length3 = 0, cnt = 0;
    int flag = 0;

    struct ListNode *tail;
    struct ListNode *l3, *l4;
    // 数据的低位在数组的低位
    while (l1 != NULL)
    {
        value1[++length1] = l1->val;
        l1 = l1->next;
    }

    while (l2 != NULL)
    {
        value2[++length2] = l2->val;
        l2 = l2->next;
    }

    while(cnt <= length1 || cnt <= length2 || flag == 1)
    {
        // 假如两个数相加大于9，则需要进一
        if(value1[cnt] + value2[cnt] + flag > 9)
        {
            value3[cnt] = value1[cnt] + value2[cnt] - 10 + flag;
            flag = 1;
        }
        else
        {
            value3[cnt] = value1[cnt] + value2[cnt] + flag;
            flag = 0;
        }
        cnt++;
    }

    l3 = (struct ListNode*)malloc(sizeof(struct ListNode));
    l3->val = value3[length3++];
    l3->next = NULL;
    tail = l3;
    while (length3 < cnt)
    {
        l4 = (struct ListNode*)malloc(sizeof(struct ListNode));
        l4->val = value3[length3++];
        l4->next = NULL;
        tail->next = l4;
        tail = l4;
    }

    return l3;
}

File: 2.两数相加/TowNumberAdd.py
class Solution:
    def addTwoNumbers(self, l1, l2):
        value1 = Solution().GenerateValue(l1)
        value2 = Solution().GenerateValue(l2)

        print(value1, value2)

        addValue = value1 + value2
        print(addValue)

        cnt = 10
        node = Node(int(addValue%cnt))
        list3  = List(node)
        addValue = int(addValue/10)
        while addValue > 0:
            node = Node(int(addValue%cnt))
            list3.addNode(node) 
            addValue = int(addValue / 10)

        val = Solution().GenerateValue(list3)
        print(val)


    def GenerateValue(self, ListNode):
        node = ListNode.val
        value = 0
        cnt = 1

        while node is not None:
            value = value + int(node.val)*cnt
            node = node.next
            cnt = cnt*10
        # print(value)
        return value

class Node:
    def __init__(self, x):
        self.val = x
        self.next = None

class List:
    def __init__(self, node):
        self.val = node
        self.next = None

    def view(self):
        node = self.val
        linkStr = ''

        while node is not None:
            linkStr = linkStr + str(node.val)
            node = node.next

        #print(linkStr)

    def addNode(self, node):
        self.next = node
        self.next.next = None


if __name__ == "__main__":
    node1 = Node(4)
    node2 = Node(5)
    node3 = Node(3)
    node4 = Node(7)
    node5 = Node(3)

    list1 = List(node1)
    list1.addNode(node2)

    list2 = List(node4)
    list2.addNode(node5)

    Solution().addTwoNumbers(list1,list2)

